src/Microsoft.ML.Tokenizers/Model/SentencePieceTokenizer.cs

```diff
@@ -1783,517 +429,33 @@ public override OperationStatus Decode(IEnumerable<int> ids, Span<char> destinat
-
-            for (int i = 1; i < symbolIndex; i++)
-            {
-                TryMerge(i - 1, i, text);
-            }
-
-            while (agenda.Count > 0)
-            {
-                SymbolPair top = agenda.Dequeue();
-
-                if (symbols[top.Left].pieceSpan.Length == 0 || symbols[top.Right].pieceSpan.Length == 0 ||
-                    symbols[top.Left].pieceSpan.Length + symbols[top.Right].pieceSpan.Length != top.Length)
-                {
-                    continue;
-                }
-
-                // Replaces symbols with `top` rule.
-                symbols[top.Left].pieceSpan = (symbols[top.Left].pieceSpan.Index, symbols[top.Left].pieceSpan.Length + symbols[top.Right].pieceSpan.Length);
-                symbols[top.Left].id = top.Id;
-
-                // Updates prev/next pointers.
-                symbols[top.Left].next = symbols[top.Right].next;
-
-                if (symbols[top.Right].next >= 0)
-                {
-                    symbols[symbols[top.Right].next].prev = top.Left;
-                }
-                symbols[top.Right].pieceSpan = (0, 0);
-
-                // Adds new symbol pairs which are newly added after symbol replacement.
-                TryMerge(symbols[top.Left].prev, top.Left, text);
-                TryMerge(top.Left, symbols[top.Left].next, text);
-            }
-
-            return revMerge;
-
-            void TryMerge(int left, int right, ReadOnlySpan<char> textSpan)
-            {
-                if (left == -1 || right == -1)
-                {
-                    return;
-                }
-
-                int pieceLength = symbols[left].pieceSpan.Length + symbols[right].pieceSpan.Length;
-                if (!_vocab.TryGetValue(textSpan.Slice(symbols[left].pieceSpan.Index, pieceLength), out (int Id, float Score, byte Type) leftId))
-                {
-                    return;
-                }
-
-                symbols[left].type = leftId.Type;
-
-                SymbolPair pair = new(left, right, leftId.Score, pieceLength, leftId.Id);
-                agenda.Enqueue(pair);
-
-                if (leftId.Type == (byte)ModelProto.Types.SentencePiece.Types.Type.Unused)
-                {
-                    revMerge ??= new();
-                    revMerge.Add((symbols[left].pieceSpan.Index, pieceLength), (symbols[left].pieceSpan.Index, symbols[left].pieceSpan.Length, symbols[right].pieceSpan.Index, symbols[right].pieceSpan.Length));
-                }
-            }
-        }
-
-        private struct SymbolPair : IEquatable<SymbolPair>, IComparable<SymbolPair>
+        /// <summary>
+        /// Creates an instance of SentencePieceTokenizer. The model stream should contain a SentencePiece model as specified in the following documentation:
+        /// https://github.com/google/sentencepiece/blob/master/src/sentencepiece_model.proto.
+        /// </summary>
+        /// <param name="modelStream">The stream containing the SentencePiece Bpe model.</param>
```

Comment:
I think this could now be a Bpe or Unigram model.
```suggestion
        /// <param name="modelStream">The stream containing the SentencePiece Bpe or Unigram model.</param>
```